autocrit.py

```diff
@@ -0,0 +1,158 @@
+
+        except Exception as e:
+            print(e)
+            time.sleep(10)
+
+    raise Exception(f"Failed after {MAX_API_RETRY} retries.")
+
+def revise(prompts, get_answer, get_critique, constitution=[], max_iterations=3, score_fn=None):
+    accelerator = Accelerator()
+    dataloader = accelerator.prepare(DataLoader(prompts))
+    outputs = []
+
+    stop = ["\nREVISION REQUEST:", "\nCRITIQUE REQUEST:", "\nUSER:", "\nASSISTANT:", "\nREVISION:", "\nCRITIQUE:"]
+    get_answer = truncate_output(get_answer, stop)
+    get_critique = truncate_output(get_critique, stop)
+
+    for (question,) in tqdm(dataloader, disable=not accelerator.is_main_process):
+        iterations = []
+        context = f"USER: {question}"
+        print0(context)
+
+        if constitution:
+            constitution_iter = islice(cycle(map(lambda x: x.values(), constitution)), max_iterations)
+
+        for i in range(max_iterations):
+            role = "ASSISTANT" if i == 0 else "REVISION"
+            context += f"\n{role}: "
+            answer = get_answer(context)
+            context += answer
+            print0(f"{role}: {answer}", style="bold")
+
+            if constitution:
+                critique_request, revision_request = next(constitution_iter)
+                context += f"\nCRITIQUE REQUEST: {critique_request}"
+                print0(f"CRITIQUE REQUEST: {critique_request}")
+
+            context += "\nCRITIQUE: "
+            critique = get_critique(context)
+            context += critique
+            print0(f"CRITIQUE: {critique}", style="italic")
+
+            if constitution:
+                context += f"\nREVISION REQUEST: {revision_request}"
+                print0(f"REVISION REQUEST: {revision_request}")
+
+            if score_fn:
+                score = score_fn(question=question, answer=answer)
+                print0(f"SCORE: {score}", style="blue")
+            else:
+                score = None
+
+            iterations.append((answer, critique, score, context))
+
+        outputs.append({
+            "question": question,
+            "iterations": [{"answer": answer, "critique": critique, "score": score, "context": context} for answer, critique, score, context in iterations]
+        })
+
+    return gather(outputs, total_size=len(prompts))
+
+def gather(samples, total_size: int):
+    if not torch.distributed.is_initialized():
+        return samples
+
+    all_samples = [None for _ in range(torch.distributed.get_world_size())]
+    torch.distributed.all_gather_object(all_samples, samples)
+    return sum(all_samples, [])[:total_size]
+
+def truncate_output(generate_fn, stop=[]):
+    def fn(*args, **kwargs):
+        output = generate_fn(*args, **kwargs)
+
+        for s in stop:
+            if s in output:
+                output = output[:output.index(s)]
+        return output
+    return fn
+
+def finetune(accelerator, model, tokenizer, optim, prompts, outputs, eval_prompts=[]):
```

Comment:
Very compact training code :) Do you think it makes sense to include this in autocrit, or should we add this as an example to trlX, with autocrit as a dependency?